From 2544c87638bc08a470aa78a7bdbffed056c28277 Mon Sep 17 00:00:00 2001
From: Orlando Vazquez <ovazquez@gmail.com>
Date: Fri, 2 Feb 2018 11:32:07 -0800
Subject: [PATCH] DOCKER-1140 Handle CNAPI error indicating docker cp'ing is
 not supported on a stopped container Reviewed by: Josh Wilsdon
 <josh@wilsdon.ca>

---
 lib/errors.js | 22 +++++++++++++++++++++-
 package.json  |  2 +-
 2 files changed, 22 insertions(+), 2 deletions(-)

diff --git a/lib/errors.js b/lib/errors.js
index 7d2d574..e3d2370 100644
--- a/lib/errors.js
+++ b/lib/errors.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2016, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 /*
@@ -566,6 +566,21 @@ PathNotDirectoryError.statusCode = 404;
 PathNotDirectoryError.description = 'path was not to a directory';
 
 
+function NotSupportedError(cause, message) {
+    _DockerBaseError.call(this, {
+        restCode: this.constructor.restCode,
+        statusCode: (cause && cause.statusCode) || this.constructor.statusCode,
+        message: message || 'operation not supported',
+        cause: cause
+    });
+}
+util.inherits(NotSupportedError, _DockerBaseError);
+NotSupportedError.prototype.name = 'NotSupportedError';
+NotSupportedError.restCode = 'NotSupported';
+NotSupportedError.statusCode = 409;
+NotSupportedError.description = 'operation not supported';
+
+
 // ---- Volume specific errors
 
 function VolumeSizeNotAvailableError(availableSizes) {
@@ -658,6 +673,10 @@ function cnapiErrorWrap(cause, message, extra) {
         case 'PathNotDirectory':
             return new PathNotDirectoryError();
 
+        case 'DockerCopyStoppedContainerNoPlatformSupport':
+            return new NotSupportedError(cause,
+                'This container must be running in order to use `docker cp`');
+
         case 'VmNotRunning':
             if (extra && extra.id) {
                 message = 'Container ' + extra.id + ' is not running';
@@ -800,6 +819,7 @@ module.exports = {
     FileNotFoundError: FileNotFoundError,
     NetworkNotFoundError: NetworkNotFoundError,
     PathNotDirectoryError: PathNotDirectoryError,
+    NotSupportedError: NotSupportedError,
 
     VolumeSizeNotAvailableError: VolumeSizeNotAvailableError,
     VolumesNotReachableError: VolumesNotReachableError,
diff --git a/package.json b/package.json
index a809285..f46eb58 100644
--- a/package.json
+++ b/package.json
@@ -1,6 +1,6 @@
 {
   "name": "sdc-docker",
-  "version": "0.4.11",
+  "version": "0.4.12",
   "author": "Joyent (joyent.com)",
   "private": true,
   "dependencies": {
-- 
2.21.0

